### PR TITLE
First set of NGFF optimizations

### PR DESCRIPTION
**Pull request by chris-allan**
Optimizations for the NGFF image and thumbnail rendering workflows consisting of:

 * Tracing of the PixelBuffer and resolution description retrieval code paths
 * Optimizing away a large number of Zarr array openings from the resolution description code path

### CODE REVIEW

**Comment by chris-allan** [comment]
Currently a work in progress. There will likely be a few more to come.

File: src/main/java/com/glencoesoftware/omero/ms/image/region/ZarrPixelBuffer.java [hunk 1]
```diff
@@ -669,11 +656,16 @@ public Dimension getTileSize() {
...
-            int[][] shapes = getShapes();
+            int resolutionLevels = getResolutionLevels();
             List<List<Integer>> resolutionDescriptions =
                     new ArrayList<List<Integer>>();
-            for (int[] shape : shapes) {
-                resolutionDescriptions.add(Arrays.asList(shape[4], shape[3]));
+            int sizeX = pixels.getSizeX();
+            int sizeY = pixels.getSizeY();
+            for (int i = 0; i < resolutionLevels; i++) {
+                double scale = Math.pow(2, i);
```
Comment (kkoz) [inline src/main/java/com/glencoesoftware/omero/ms/image/region/ZarrPixelBuffer.java:667]: Are the resolution levels always successive powers of 2 from each other?

**Review by kkoz — COMMENTED** [review]
I don't really have any objection to the `getBytes` changes, but I'm not sure they will prevent any copying from taking place. I think once we call `array.read(...)` the data is put in memory and then is just passed around.

File: src/main/java/com/glencoesoftware/omero/ms/image/region/ZarrPixelBuffer.java (hunk 1 above)
Comment (chris-allan) [reply src/main/java/com/glencoesoftware/omero/ms/image/region/ZarrPixelBuffer.java:667]: Yes, at least as far as we use them. It's not explicitly defined but should be in the future:

* https://github.com/zarr-developers/zarr-specs/issues/50

**Comment by chris-allan** [comment]
Depending on the exact implementation I was concerned about `ByteBuffer#array()`. In our use currently I don't think it's a problem and where I'd definitely like to go with all of this is to use preallocated arrays and `ZarrArray.read(Object, int[], int[]` where we are passing in a preallocated array but that's probably for `002`. ;)

**Pull request merged**